addons/pos_online_payment/tests/test_frontend.py

```diff
@@ -0,0 +1,216 @@
+            'invoice_journal_id': cls.sales_journal.id,
+            'journal_id': cls.sales_journal.id,
+            'payment_method_ids': [Command.link(cls.cash_payment_method.id), Command.link(cls.online_payment_method.id)],
+        })
+
+        # Code from addons/point_of_sale/tests/test_frontend.py:
+        cls.pos_user = cls.env['res.users'].create({
+            'name': 'A simple PoS man!',
+            'login': 'pos_op_user',
+            'password': 'pos_op_user',
+            'groups_id': [
+                (4, cls.env.ref('base.group_user').id),
+                (4, cls.env.ref('point_of_sale.group_pos_user').id),
+            ],
+        })
+        cls.pos_user.partner_id.email = 'pos_op_user@test.com'
+        # End of code from addons/point_of_sale/tests/test_frontend.py
+
+    # Code from addons/account_payment/tests/common.py
+    @classmethod
+    def _prepare_provider(cls, provider_code='none', company=None, update_values=None):
+        """ Override of `payment` to prepare and return the first provider matching the given
+        provider and company.
+
+        If no provider is found in the given company, we duplicate the one from the base company.
+        All other providers belonging to the same company are disabled to avoid any interferences.
+
+        :param str provider_code: The code of the provider to prepare.
+        :param recordset company: The company of the provider to prepare, as a `res.company` record.
+        :param dict update_values: The values used to update the provider.
+        :return: The provider to prepare, if found.
+        :rtype: recordset of `payment.provider`
+        """
+        provider = super()._prepare_provider(provider_code, company, update_values)
+        if not provider.journal_id:
+            provider.journal_id = cls.env['account.journal'].search(
+                [('company_id', '=', provider.company_id.id), ('type', '=', 'bank')],
+                limit=1,
+            )
+        return provider
+    # End of code from addons/account_payment/tests/common.py
+
+    def setUp(self):
+        self.enable_reconcile_after_done_patcher = False
+
+        real_get_and_set_online_payments_data = PosOrder.get_and_set_online_payments_data
+
+        def _fake_get_and_set_online_payments_data(method_self, next_online_payment_amount=False):
+            pos_order_id = method_self.id
+            order_access_token = method_self.access_token
+            expected_payment_provider_id = self.payment_provider.id
+            if not isinstance(next_online_payment_amount, bool) and not tools.float_is_zero(next_online_payment_amount, precision_rounding=method_self.currency_id.rounding) and next_online_payment_amount > 0:
+                # Delay must be long enough to execute _fake_online_payment after the current RPC call is answered
+                t = threading.Timer(5, self._fake_online_payment, args=(pos_order_id, order_access_token, expected_payment_provider_id))
```

Comment:
This is a bad idea to spawn a thread inside a test, it is a very good way to introduce randomness. 

The request that are triggered inside _fake_online_payment should have a test cursor and be sequential with the other request, but any call to self.env.cr will be concurrent and can lead to concurrent access to the cursor. 
 
 As an example _get_tx will make a search, so will execute and fetch on the same cr as a request, and concurrently.
 
 This can lead to error like in https://runbot.odoo.com/runbot/build/54372105
 
 ```
   File "/data/build/odoo/odoo/models.py", line 3597, in _fetch_query
    rows = self.env.cr.fetchall()
psycopg2.ProgrammingError: no results to fetch
```
This should be done sequentially. It is hard to do since inside a request, the thread will deadlock trying to generate another request. This is why the timout is needed here.

A solution would be to split the tour in parts, start the tour, at the end, simulate the answer, then start the rest of the tour. 
Another solution is to simulate the answer without making a request, by calling the code inside the endpoint directly 
A last proposition is to trigger the endpoint response from the tour if it makes sense.

Reply:
Hello @Xavier-Do, thank you for your explanations and your suggestions.

The idea of this test is to simulate an almost real flow: the cashier initiates an online payment, the server receives it and answers normally to the cashier, but the server also uses the received request to generate a fake online payment (by doing web requests to the online payment page, as a normal customer would do to pay). The cashier then receives this fake online payment and checks it.
(Because websocket communication doesn't seem to work in the test environment, I had to make the JS tour web client frequently request the state of the (fake) online payment from the server)

I'm not sure to understand what you wrote, could you confirm me the following please ?

> The request that are triggered inside _fake_online_payment

By "request", I guess you are talking about _fake_open_pos_order_pay_page, _fake_request_pos_order_pay_transaction_page and _fake_open_pos_order_pay_confirmation_page.

> should have a test cursor and be sequential with the other request

Are you describing the current code ("should" used for an assumption) or are you asking for that to be done ("should" used to express a condition to follow to make the test correctly work) ?
By "other request", are you talking about _fake_get_and_set_online_payments_data ?

> but any call to self.env.cr will be concurrent and can lead to concurrent access to the cursor.

Any call from where ? During the execution of the thread for _fake_online_payment ?

> As an example _get_tx will make a search, so will execute and fetch on the same cr as a request, and concurrently.

Is the "request" in your sentence referring to any request or a special request in the test ?
What I understand is that since _fake_online_payment is executed in a separate thread, and because it executes _get_tx which does a search with the cursor, another thread could also do a request to the same cursor at the same time, concurrently.

> This should be done sequentially.

By "This" I assume you are talking about _fake_online_payment execution.

> It is hard to do since inside a request

By "request" I assume you are talking about the RPC to _fake_get_and_set_online_payments_data.

> the thread will deadlock trying to generate another request. This is why the timout is needed here.

I don't understand what timeout you are talking about.
I understand that the thread executing the RPC to _fake_get_and_set_online_payments_data would deadlock with the web requests in _fake_online_payment if _fake_online_payment was called synchronously in _fake_get_and_set_online_payments_data.
I think that it was my first approach, but indeed it seems that doing web communications inside an RPC is not supported.

I would like to know, is the problem caused by the test itself (depending on how the 2 threads (the one for the RPC to _fake_get_and_set_online_payments_data and the one for executing _fake_online_payment) of the test are executed (slow/fast execution because of busy test server for example ?), they can concurrently use the same cursor ?), or the way it is executed (are several tests executed simultaneously for example, and inducing the random error ?).

> A solution would be to split the tour in parts, start the tour, at the end, simulate the answer, then start the rest of the tour.

What I understand is that you suggest to have 2 tours, and between those 2 tours simulate the answer.
From my understanding, this is not possible and wouldn't represent a realistic flow because the web client would lose the online payment popup (showing the current online payment, and waiting for the successful online payment notification from the server). It wouldn't be possible to open a second tour with the state of the online payment popup displayed and all the JS code ready to receive the answer correctly (of course everything is possible, but from my point of view it would need a lot of hacking code specific to the test, that at the end wouldn't test the actual working of pos_online_payment).

> Another solution is to simulate the answer without making a request, by calling the code inside the endpoint directly

I understand that you suggest to directly call pos_order_pay, pos_order_pay_transaction and pos_order_pay_confirmation of payment_portal.py from the _fake_online_payment implementation (and call _fake_online_payment synchronously in _fake_get_and_set_online_payments_data).
How would the "request" variable be filled ? Wouldn't it cause issues to execute controller methods like that (from the execution of the RPC to _fake_get_and_set_online_payments_data) ?

> A last proposition is to trigger the endpoint response from the tour if it makes sense.

I'm sorry I don't understand. That's already what is done: the tour is triggering the server to do the fake requests to the online payment page (with the pos_order_id and order_access_token that are generated during the tour and are received from it).
The idea of the test is to test a full flow: the cashier interface and the customer payment web interface. If the web requests to the payment page are removed, the test would become significantly less interesting.

I would have a last question please: Is it possible to schedule the execution of some code (_fake_online_payment) after some delay on the main thread (and not a separate one as I did) during the execution of a tour ?
If it's not possible, then I think that I will just simplify the test.
Thank you in advance for your explanations.